### PR TITLE
Feature/jc19 360 details for a selected film

### PR DESCRIPTION
**Pull request by LocTran-saritasa**
Details for a selected film.

### CODE REVIEW

File: apps/vanilla/src/script/renderDetail.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+/**
+ * Create one raw in the information of anime.
+ * @param title Title of the information.
+ * @param content Content of the information.
+ */
+function createInformationRaw(title: string, content: string): HTMLElement {
+  const label = document.createElement('label');
+  label.classList.add('detail__information-label');
+  const titleNode = document.createElement('span');
```
Comment (gevorg-gevorgyan-saritasa) [inline apps/vanilla/src/script/renderDetail.ts:44]: Maybe `titleElement`?

File: apps/vanilla/src/global/header.css [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+  background-color: var(--primary-color);
+  position: fixed;
+  top: 0;
+  left: 0;
+  display: flex;
+  justify-content: space-between;
+}
+
+.header__page,
+.header__authen {
```
Comment (panov-aleksey-saritasa) [inline apps/vanilla/src/global/header.css:13]: What is *authen*?

File: apps/vanilla/src/namespaces/PageHandler.ts [hunk 3]
```diff
@@ -29,12 +38,8 @@ export namespace PageHandler {
   export function goToLastPage(): void {
     const count = localStorage.getItem(COUNT_LS);
     assertNonNullish(count);
-    let page = Number.parseInt(count, DECIMAL) / LIMIT;
-    if (isNaN(page) === true) {
-      page = 1;
-    } else {
-      page = Math.ceil(page);
-    }
+    const totalPage = Number.parseInt(count, DECIMAL) / LIMIT;
```
Comment (panov-aleksey-saritasa) [inline apps/vanilla/src/namespaces/PageHandler.ts:41]: I think you don't need `DECIMAL` constant, it's clear what does `10` mean in context of `parsetInt` 🙂

File: apps/vanilla/src/script/constants/localStorageName.ts [hunk 4]
```diff
@@ -0,0 +1,7 @@
+/** Local storage name. */
+export const COUNT_LS = 'COUNT';
+export const SORT_LS = 'ANIME_SORT';
+export const ORDER_LS = 'ANIME_ORDER';
+export const ANIME_LS = 'ANIME_ID';
```
Comment (panov-aleksey-saritasa) [inline apps/vanilla/src/script/constants/localStorageName.ts:5]: https://wiki.saritasa.rocks/frontend/languages/ts-js/naming#avoid-contractions

File: apps/vanilla/src/services/fetchAnime.ts [hunk 5]
```diff
@@ -1,19 +1,22 @@
...
-import { Detail } from '@js-camp/core/models/detail';
 
 import { http } from '../api';
-import { ANIME_ROUTE, COUNT_LS } from '../script/constants';
+import { COUNT_LS } from '../script/constants/localStorageName';
 import { renderAnime, renderPagination } from '../script/renderToUI';
 
+/** Request address. */
+export const ANIME_ROUTE = '/api/v1/anime/anime/';
+
```
Comment (panov-aleksey-saritasa) [inline apps/vanilla/src/services/fetchAnime.ts:19]: In the context of frontend a *route* usually means navigation-related stuff. Better to use *url* here to not confuse anyone

File: apps/vanilla/src/namespaces/middleware.ts [hunk 6]
```diff
@@ -20,4 +22,24 @@ export namespace Middleware {
...
+      try {
+        const tokens = await getRefreshedToken(refreshToken);
+        storeTokens(tokens);
+        return true;
+      } catch (error: unknown) {
+        return false;
+      }
+    }
+    return true;
+  }
```
Comment (panov-aleksey-saritasa) [inline apps/vanilla/src/namespaces/middleware.ts:44]: Right now this function does two separate things at the same time: checks if there are tokens and implicitly verifies/refreshes them. Consider separating this stuff into different functions

Comment (panov-aleksey-saritasa) [reply apps/vanilla/src/namespaces/middleware.ts:44]: About doing token refresh, [axios.interceptors](https://axios-http.com/docs/interceptors) is generally a good place to do this stuff, you need to just check for [401](https://www.webfx.com/web-development/glossary/http-status-codes/what-is-a-401-status-code/) errors from axios and try to do refresh token logic

File: apps/vanilla/src/services/auth.ts [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+      localStorage.setItem(REFRESH_TOKEN, auth.refreshToken);
+      return null;
+    } catch (error: unknown) {
+      const errorData = HttpErrorMapper.fromDto(error as HttpErrorDto);
+      const errorList = [];
+      for (const i of Object.keys(errorData.data)) {
+        errorList.push(...errorData.data[i]);
+      }
+      errorList.push(errorData.detail);
+      return errorList;
```
Comment (panov-aleksey-saritasa) [inline apps/vanilla/src/services/auth.ts:58]: duplicates error catching from the `login` function, consider at least moving this out as a function

File: apps/vanilla/src/services/auth.ts [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+   */
+  export async function login(data: LoginData): Promise<Array<string> | null> {
+    try {
+      const response = await http.post<TokenDto>(ApiUrl.login, data);
+      const auth = TokenMapper.fromDto(response.data);
+      localStorage.setItem(ACCESS_TOKEN, auth.accessToken);
+      localStorage.setItem(REFRESH_TOKEN, auth.refreshToken);
+      return null;
+    } catch (error: unknown) {
+      const errorData = HttpErrorMapper.fromDto(error as HttpErrorDto);
```
Comment (panov-aleksey-saritasa) [inline apps/vanilla/src/services/auth.ts:27]: use `instanceof HttpErrorDto` check instead of typecasting

File: apps/vanilla/src/services/auth.ts [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+import { ApiUrl } from '../namespaces/apiUrl';
+import { ACCESS_TOKEN, REFRESH_TOKEN } from '../script/constants/localStorageName';
+
+export namespace Auth {
+
+  /**
+   * Handle login request.
+   * @param data Login data.
+   */
+  export async function login(data: LoginData): Promise<Array<string> | null> {
```
Comment (panov-aleksey-saritasa) [inline apps/vanilla/src/services/auth.ts:19]:
```suggestion
  export async function login(data: LoginData): Promise<string[] | null> {
```

File: apps/vanilla/src/services/auth.ts [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+    } catch (error: unknown) {
+      const errorData = HttpErrorMapper.fromDto(error as HttpErrorDto);
+      const errorList = [];
+      if (errorData.data !== undefined) {
+        for (const i of Object.keys(errorData.data)) {
+          errorList.push(...errorData.data[i]);
+        }
+      }
+      errorList.push(errorData.detail);
+      return errorList;
```
Comment (panov-aleksey-saritasa) [inline apps/vanilla/src/services/auth.ts:35]: Consider throwing `HttpError` instead

Comment (panov-aleksey-saritasa) [reply apps/vanilla/src/services/auth.ts:35]: From the signature of this function is very very unclear that `Array<string>` is actually array of error messages 🥲

**Review by panov-aleksey-saritasa — CHANGES_REQUESTED** [review]

File: apps/vanilla/src/global/header.css (hunk 2 above)
Comment (LocTran-saritasa) [reply apps/vanilla/src/global/header.css:13]: I mean authenticate. Let me fix it :sweat_smile:

File: apps/vanilla/src/services/auth.ts (hunk 10 above)
Comment (LocTran-saritasa) [reply apps/vanilla/src/services/auth.ts:35]: I want to get the error message from BE and display on UI. If I throw an error here, I can't get these messages. Is there any solution for this?

File: apps/vanilla/src/services/auth.ts (hunk 8 above)
Comment (LocTran-saritasa) [reply apps/vanilla/src/services/auth.ts]: I got this error while use instanceof HttpErrorDto
> 'HttpErrorDto' only refers to a type, but is being used as a value here. ts(2693)

File: apps/vanilla/src/api/interceptors.ts [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+    throw new Error('Axios did not pass any header.');
+  }
+  if (localStorage.getItem('ACCESS_TOKEN') !== null) {
+    headers.Authorization = `Bearer ${localStorage.getItem('ACCESS_TOKEN') ?? ''}`;
+  }
+  return {
+    ...config,
+    headers: {
+      ...headers,
+      'Api-Key': CONFIG.apiKey,
```
Comment (thanhnguyen-saritasa) [inline apps/vanilla/src/api/interceptors.ts:27]: You could use constant API_HEADER already have

File: apps/vanilla/src/namespaces/PageHandler.ts [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+/** Default values of query if user don't pass any value. */
+export const DEFAULT_ORDERING = Sorting.Default;
+
+export namespace PageHandler {
+
+  /**
+   * Get new data for next page.
+   * @param newPage Next page number.
+   */
+  export function goToPageByNum(newPage: number): void {
```
Comment (thanhnguyen-saritasa) [inline apps/vanilla/src/namespaces/PageHandler.ts:22]: Avoid contraction, please improve the funtion name `goToNextPage` is ok?
https://wiki.saritasa.rocks/frontend/languages/ts-js/naming#avoid-contractions

File: apps/vanilla/src/namespaces/middleware.ts [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+
+  /** Check if user is logged in then redirect to profile page. */
+  export function checkIsNotLoggedIn(): void {
+    const accessToken = localStorage.getItem(ACCESS_TOKEN);
+    const refreshToken = localStorage.getItem(REFRESH_TOKEN);
+    if (accessToken === null || refreshToken === null) {
+      return;
+    }
+    location.replace(PROFILE_PAGE);
+  }
```
Comment (timofei-maximov-saritasa) [inline apps/vanilla/src/namespaces/middleware.ts:23]: Let's use `isAuthorized` function to reduce code duplication

File: libs/core/models/detail.ts [hunk 14]
```diff
@@ -0,0 +1,79 @@
+import { Anime } from './anime';
+import { OmitImmerable } from './immerable';
+
+/** Anime detail data. */
+export class Detail extends Anime {
```
Comment (thanhnguyen-saritasa) [inline libs/core/models/detail.ts:5]: I think the name of file and class is common meaning.
Would you please update it to 'AnimeDetail'

File: libs/core/mappers/anime.mapper.ts [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+  /**
+   * Maps dto to model.
+   * @param dto Anime dto.
+   */
+  export function fromDto(dto: AnimeDto): Anime {
+
+    const status = animeStatusDtoToModel[dto.status];
+    const type = animeTypeDtoToModel[dto.type];
+
+    return {
```
Comment (thanhnguyen-saritasa) [inline libs/core/mappers/anime.mapper.ts:32]: Do you see the IDE warning missing property [immerable]?
Please return new Enime()

File: libs/core/dtos/detail.dto.ts [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+  readonly id: number;
+
+  /** Image of the anime. */
+  readonly image: string;
+
+  /** Link trailer on youtube. */
+  readonly trailer_youtube_id: string;
+
+  /** Title in English of the anime. */
+  readonly title_eng: string;
```
Comment (thanhnguyen-saritasa) [inline libs/core/dtos/detail.dto.ts:17]: Please avoid contraction.
https://wiki.saritasa.rocks/frontend/languages/ts-js/naming#avoid-contractions
'title_english'

File: libs/core/dtos/detail.dto.ts [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+
+  /** Title in Japanese of the anime. */
+  readonly title_jpn: string;
```
Comment (thanhnguyen-saritasa) [inline libs/core/dtos/detail.dto.ts:20]: Same above suggestion, please improve it

File: libs/core/dtos/anime.dto.ts [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+/** Define data in Anime dto. */
+export interface AnimeDto {
+
+  /** Id of the anime. */
+  readonly id: number;
+
+  /** English title of the anime. */
+  readonly title_eng: string;
```
Comment (thanhnguyen-saritasa) [inline libs/core/dtos/anime.dto.ts:27]: please update property name
https://wiki.saritasa.rocks/frontend/languages/ts-js/naming#avoid-contractions

File: apps/vanilla/src/services/token.ts [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+
+/**
+ * Return the refreshed token.
+ * @param refreshToken Refresh token.
+ */
+export async function getRefreshedToken(refreshToken: string): Promise<Token> {
+  const response = await http.post<TokenDto>(ApiUrl.refresh, {
+    token: refreshToken,
+  });
+  return TokenMapper.fromDto(response.data);
```
Comment (thanhnguyen-saritasa) [inline apps/vanilla/src/services/token.ts:32]: What will it happenning when API request failed? I think we could handle it in try ... catch

File: libs/core/dtos/detail.dto.ts (hunk 16 above)
Comment (LocTran-saritasa) [reply libs/core/dtos/detail.dto.ts]: Because this is dto interfaces so I had to name it like that due to the data response from BE.

**Pull request closed without merging**